src/compiler/scala/tools/nsc/ast/parser/Parsers.scala

```diff
@@ -810,9 +810,22 @@ self =>
       makeTupleTerm(elems)
     }
 
+    /** Create a function Tree. If the arity is not supported, a syntax error is emitted. */
+    def makeSafeFunctionType(argtpes: List[Tree], offset: Offset, restpe: Tree) = {
+      if (checkFunctionArity(argtpes, offset)) makeFunctionTypeTree(argtpes, restpe)
+      else makeFunctionTypeTree(Nil, restpe) // create a dummy node
+    }
+
     private[this] def checkTupleSize(elems: List[Tree], offset: Offset): Boolean =
       elems.lengthCompare(definitions.MaxTupleArity) <= 0 || {
-        val msg = s"too many elements for tuple: ${elems.length}, allowed: ${definitions.MaxTupleArity}"
+        val msg = s"tuples may not have more than ${definitions.MaxFunctionArity} elements, but ${elems.length} given"
+        syntaxError(offset, msg, skipIt = false)
+        false
+      }
+
+    private[this] def checkFunctionArity(argtpes: List[Tree], offset: Offset): Boolean =
+      argtpes.lengthCompare(definitions.MaxFunctionArity) <= 0 || {
+        val msg = s"function values may not have more than ${definitions.MaxFunctionArity} parameters, but ${argtpes.length} given"
         syntaxError(offset, msg, skipIt = false)
         false
```

Comment:
on the topic of error messages, what if we reported the syntax error at the `pos` of the `MaxFunctionArity`th param/arg?

Reply:
I like this idea a lot, that might take me a bit of time to figure out but I think it would be worth it.